lib/SILOptimizer/Mandatory/TFConstExpr.cpp

```diff
@@ -996,146 +1010,66 @@ SymbolicValue ConstExprFunctionState::getConstantValue(SILValue value) {
-  }
-
-  unsigned elementNo = indices.front();
-
-  // If we have a non-aggregate then fail.
-  if (aggregate.getKind() != SymbolicValue::Aggregate &&
-      aggregate.getKind() != SymbolicValue::Array)
-    return true;
-
-  ArrayRef<SymbolicValue> oldElts;
-  Type eltType;
-
-  // We need to have an array, struct or a tuple type.
-  if (aggregate.getKind() == SymbolicValue::Array) {
-    CanType arrayEltTy;
-    oldElts = aggregate.getArrayValue(arrayEltTy);
-    eltType = arrayEltTy;
-  } else {
-    oldElts = aggregate.getAggregateValue();
-
-    if (auto *decl = type->getStructOrBoundGenericStruct()) {
-      auto it = decl->getStoredProperties().begin();
-      std::advance(it, elementNo);
-      eltType = (*it)->getType();
-    } else if (auto tuple = type->getAs<TupleType>()) {
-      assert(elementNo < tuple->getNumElements() && "invalid index");
-      eltType = tuple->getElement(elementNo).getType();
-    } else {
-      return true;
-    }
-  }
-
-  if (writeOnlyOnce &&
-      oldElts[elementNo].getKind() != SymbolicValue::UninitMemory) {
-    // Cannot overwrite an existing constant.
-    return true;
-  }
-
-  // Update the indexed element of the aggregate.
-  SmallVector<SymbolicValue, 4> newElts(oldElts.begin(), oldElts.end());
-  if (updateIndexedElement(newElts[elementNo], indices.drop_front(), scalar,
-                           eltType, writeOnlyOnce, allocator))
-    return true;
-
-  if (aggregate.getKind() == SymbolicValue::Aggregate)
-    aggregate = SymbolicValue::getAggregate(newElts, allocator);
-  else
-    aggregate = SymbolicValue::getArray(newElts, eltType->getCanonicalType(),
-                                        allocator);
-  return false;
-}
-
-/// Find the initializer (single writer) of `addr` among it users,
-/// const-evaluate it and store the result into a memory object, and return the
-/// address of that memory object.
+/// This is a helper function for `getSingleWriterAddressValue`. Callers should
+/// use `getSingleWriterAddressValue`.
 ///
-/// Some use cases are:
-/// 1. When analyzing the top-level code involved in a constant expression, we
-/// can end up demanding values that are returned by address.  Handle this by
-/// finding the temporary stack value (an alloc_stack inst), and calling this
-/// method on it.
-/// 2. When const-evaluating an array via decodeAllocUninitializedArray(),
-/// do that by const-evaluating the writers of individual array elements.
+/// If `addr` has no writing uses, returns {false, None} and has no side
+/// effects.
 ///
-///  There are a few forms of writers, such as:
-/// - store %3 to %4 ...
-/// - %8 = pointer_to_address %7 : $Builtin.RawPointer to [strict] $*Int32
-/// - %14 = index_addr %9 : $*Int32, %13 : $Builtin.Word
-/// - %180 = tuple_element_addr %179 : $*(Int32, Int32, Int32, Int32), 3
+/// If the following conditions hold:
+///   * `addr` points at uninitialized memory;
+///   * there are store(s) to `addr` that, taken together, set the memory
```

Comment:
should we rename "store" to "write" or sth, to clarify there are different writer inst types?

Reply:
yep